build-info-extractor-pip/src/main/java/org/jfrog/build/extractor/pip/extractor/PipBuildInfoExtractor.java

```diff
@@ -57,58 +59,126 @@ Build extract(ArtifactoryDependenciesClient client, String repository, String in
-            Dependency dependency = null;
             if (StringUtils.isNotBlank(fileName)) {
-                // Get dependency info from Artifactory - may throw IOException.
-                dependency = getPackageInfoFromArtifactory(fileName, repository, client, logger);
-            } else {
-                // Get dependency info from cache.
-                if (dependenciesCache != null) {
-                    dependency = dependenciesCache.getDependency(pkgName);
-                }
-            }
-
-            if (dependency == null) {
-                // Add missing dependency.
-                missingDeps.add(pkgName);
+                // Dependency downloaded from Artifactory.
+                getFromArtifactoryMap.put(fileName, pkgName);
                 continue;
             }
+            // Dependency wasn't downloaded in this execution, get dependency info from cache.
+            if (dependenciesCache != null && dependenciesCache.getDependency(pkgName) != null) {
+                dependenciesMap.put(pkgName, dependenciesCache.getDependency(pkgName));
+            }
+        }
 
-            dependenciesMap.put(pkgName, dependency);
+        // Get dependencies from Artifactory.
+        Map<String, Dependency> dependenciesFromArtifactory = getDependenciesFromArtifactory(getFromArtifactoryMap, repository, client, logger);
+        if (dependenciesFromArtifactory != null) {
+            dependenciesMap.putAll(dependenciesFromArtifactory);
         }
 
+        // Prompt missing dependencies.
+        Set<String> missingDeps = downloadedDependencies.keySet().stream()
+                .filter(x -> !dependenciesMap.containsKey(x))
+                .collect(Collectors.toSet());
         promptMissingDeps(missingDeps, logger);
+
         DependenciesCache.updateDependenciesCache(dependenciesMap, executionPath);
         return dependenciesMap;
     }
 
-    Dependency getPackageInfoFromArtifactory(String fileName, String repository, ArtifactoryDependenciesClient client, Log logger) throws IOException {
-        String aql = String.format(PIP_AQL_FORMAT, repository, fileName);
-        AqlSearchResult searchResult;
+    /**
+     * Fetch required files-data from Artifactory and create its corresponding Dependency object.
+     * @param fileToPackageMap - Mapping between a downloaded file to its package name.
+     * @param repository - Resolution repository.
+     * @param client - Artifactory client for fetching artifacts data.
+     * @param logger - The logger.
+     * @return Mapping of a package-name and its Dependency object.
+     * @throws IOException
+     */
+    private Map<String, Dependency> getDependenciesFromArtifactory(Map<String, String> fileToPackageMap, String repository,
+                                                                   ArtifactoryDependenciesClient client, Log logger) throws IOException {
+        if (fileToPackageMap.isEmpty()) {
```

Comment:
It may be easier not to return null here, but instead return an empty map, and by that, avoid the null check right after this method is invoked.